include/ginkgo/core/preconditioner/isai.hpp

```diff
@@ -0,0 +1,280 @@
+
+
+/**
+ * The Incomplete Sparse Approximate Inverse (ISAI) Preconditioner generates
+ * approximate inverse matrices for a given lower triangular matrix L and upper
+ * triangular matrix U. Using the preconditioner computes $aiU * aiL * x$
+ * for a given vector x (may have multiple right hand sides). aiU and aiL
+ * are the approximate inverses for U and L respectively.
+ *
+ * The sparsity pattern used for the approximate inverses is the same as
+ * the sparsity pattern of the respective triangular matrix.
+ * The L and U matrices need to be in a Composition<ValueType> object in the
+ * order: L, U.
+ *
+ * For more details on the algorithm, see the paper
+ * <a href="https://doi.org/10.1016/j.parco.2017.10.003">
+ * Incomplete Sparse Approximate Inverses for Parallel Preconditioning</a>,
+ * which is the basis for this work.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ * @ingroup isai
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Isai : public EnableLinOp<Isai<ValueType, IndexType>> {
+    friend class EnableLinOp<Isai>;
+    friend class EnablePolymorphicObject<Isai, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using Csr = matrix::Csr<ValueType, IndexType>;
+
+    /**
+     * Returns the approximate inverse of L.
+     *
+     * @returns the approximate inverse of L
+     */
+    std::shared_ptr<const Csr> get_approx_inverse_l() const { return l_inv_; }
+
+    /**
+     * Returns the approximate inverse of U.
+     *
+     * @returns the approximate inverse of U
+     */
+    std::shared_ptr<const Csr> get_approx_inverse_u() const { return u_inv_; }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * When set to true, the Isai preconditioner expects a lower triangular
+         * factor (only, so no U factor), and will only generate the
+         * approximate inverse for that L factor.
+         *
+         * @note: Setting both exclusive_factor_l and exclusive_factor_u
+         *        results in an exception being thrown when generating the
+         *        preconditioner.
+         */
+        bool GKO_FACTORY_PARAMETER(exclusive_factor_l, false);
+
+        /**
+         * When set to true, the Isai preconditioner expects an upper triangular
+         * factor (only, so no L factor), and will only generate the
+         * approximate inverse for that U factor.
+         *
+         * @note: Setting both exclusive_factor_l and exclusive_factor_u
+         *        results in an exception being thrown when generating the
+         *        preconditioner.
+         */
+        bool GKO_FACTORY_PARAMETER(exclusive_factor_u, false);
+
+        /**
+         * @brief Optimization parameter that skips the sorting if the input
+         *        matrix/matrices (required by the algorithm).
```

Comment:
```suggestion
         *        matrix/matrices are sorted (required by the algorithm).
```
Maybe miss something here?